### PR TITLE
out_kafka: Introduce raw_log_key to write a single value to kafka

### PR DESCRIPTION
**Pull request by zecke**
Implement a new option called raw_log_key that allows to write a single value to kafka. This allows to use message_key to put some fields into the Kafka message_key and another field as the payload. Make this work as an additional format.

----
Enter `[N/A]` in the box, if an item is not applicable to your change.

**Testing**
Before we can approve your change; please submit the following in a comment:

- [x] Example configuration file for the change
- [x] Debug log output from testing the change

- [x] Attached [Valgrind](https://valgrind.org/docs/manual/quick-start.html) output that shows no leaks or memory corruption was found

If this is a change to packaging of containers or native binaries then please confirm it works for all targets.

- [N/A] Run [local packaging test](./packaging/local-build-all.sh) showing all targets (including any new ones) build.
- [N/A] Set `ok-package-test` label to test for all targets (requires maintainer to do).

**Documentation**

- [x] Documentation required for this feature

**Backporting**

- [ ] Backport to latest stable release.

----

Fluent Bit is licensed under Apache 2.0, by submitting this pull request I understand that this code will be released under the terms of that license.

### CODE REVIEW

**Comment by zecke** [comment]
Example config used for testing this:

```
[INPUT]
    Name  random

[FILTER]
    Name type_converter
    Match *
    uint_key rand_value rand_value string

[FILTER]
    Name modify
    Match *
    Add message Foo123

[OUTPUT]
    Name        kafka
    Match       *
    Brokers     127.0.0.1:9092
    Topics      test-foo
    Format      raw
    raw_log_key rand_value
    message_key message
```

Using kcat to read from kafka:

```
$ kcat -C -b localhost -t test-foo
% Reached end of topic test-foo [0] at offset 470
12174776462032818091
% Reached end of topic test-foo [0] at offset 471
3397822772916645746
% Reached end of topic test-foo [0] at offset 472
3284960532694686745
% Reached end of topic test-foo [0] at offset 473
18034299562221828230
% Reached end of topic test-foo [0] at offset 474
2331075259719154332
```

**Comment by zecke** [comment]
Fixed an unrelated issue found by valgrind (and added a dedicated commit). The reachable bits belong to global state created by curl (and then two dependencies down...)

```
==4403== 
==4403== HEAP SUMMARY:
==4403==     in use at exit: 192 bytes in 12 blocks
==4403==   total heap usage: 7,620 allocs, 7,608 frees, 5,391,102 bytes allocated
==4403== 
==4403== 48 bytes in 6 blocks are still reachable in loss record 1 of 2
==4403==    at 0x48850C8: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-arm64-linux.so)
==4403==    by 0x576F83F: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x5770E2F: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x57D7C5B: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x576F73F: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x57708DB: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x576C443: gcry_control (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x50F2BC3: libssh2_init (in /usr/lib/aarch64-linux-gnu/libssh2.so.1.0.1)
==4403==    by 0x4DDF8CF: ??? (in /usr/lib/aarch64-linux-gnu/libcurl.so.4.7.0)
==4403==    by 0x4D8D1FB: curl_global_init (in /usr/lib/aarch64-linux-gnu/libcurl.so.4.7.0)
==4403==    by 0x104FF87: rd_http_global_init (rdhttp.c:443)
==4403==    by 0xEC4177: rd_kafka_global_init0 (rdkafka.c:160)
==4403== 
==4403== 144 bytes in 6 blocks are still reachable in loss record 2 of 2
==4403==    at 0x48850C8: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-arm64-linux.so)
==4403==    by 0x576F83F: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x5770E2F: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x57D7C4F: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x576F73F: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x57708DB: ??? (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x576C443: gcry_control (in /usr/lib/aarch64-linux-gnu/libgcrypt.so.20.2.8)
==4403==    by 0x50F2BC3: libssh2_init (in /usr/lib/aarch64-linux-gnu/libssh2.so.1.0.1)
==4403==    by 0x4DDF8CF: ??? (in /usr/lib/aarch64-linux-gnu/libcurl.so.4.7.0)
==4403==    by 0x4D8D1FB: curl_global_init (in /usr/lib/aarch64-linux-gnu/libcurl.so.4.7.0)
==4403==    by 0x104FF87: rd_http_global_init (rdhttp.c:443)
==4403==    by 0xEC4177: rd_kafka_global_init0 (rdkafka.c:160)
==4403== 
==4403== LEAK SUMMARY:
==4403==    definitely lost: 0 bytes in 0 blocks
==4403==    indirectly lost: 0 bytes in 0 blocks
==4403==      possibly lost: 0 bytes in 0 blocks
==4403==    still reachable: 192 bytes in 12 blocks
==4403==         suppressed: 0 bytes in 0 blocks
```

**Comment by zecke** [comment]
I have rebased this against the master branch.

**Comment by agup006** [comment]
@cosmo0920  would you have some time to take a look

**Comment by patrick-stephens** [comment]
@zecke can you link the docs PR for this as well? It's a new parameter I believe is what you're saying so should be documented here (ideally with an example): https://github.com/fluent/fluent-bit-docs/blob/master/pipeline/outputs/kafka.md

I see you've ticked the backport option as well, that typically means also linking a PR to the branch for the backport version.

**Comment by zecke** [comment]
> @zecke can you link the docs PR for this as well? It's a new parameter I believe is what you're saying so should be documented here (ideally with an example): https://github.com/fluent/fluent-bit-docs/blob/master/pipeline/outputs/kafka.md

I opened https://github.com/fluent/fluent-bit-docs/pull/1397 Not sure how to link these two together.

> I see you've ticked the backport option as well, that typically means also linking a PR to the branch for the backport version.

I have unticked the box. Getting this into any future release will be great.

File: plugins/out_kafka/kafka_config.c [hunk 1]
```diff
@@ -228,10 +239,6 @@ int flb_out_kafka_destroy(struct flb_out_kafka *ctx)
         flb_free(ctx->topic_key);
     }
 
-    if (ctx->message_key) {
```
Comment (cosmo0920) [inline plugins/out_kafka/kafka_config.c:231]: :+1:

**Review by cosmo0920 — APPROVED** [review]
Looks good. Thank you!

**Comment by zecke** [comment]
The doc PR has been merged (https://github.com/fluent/fluent-bit-docs/pull/1397)

File: plugins/out_kafka/kafka.c [hunk 2]
```diff
@@ -211,6 +213,15 @@ int produce_message(struct flb_time *tm, msgpack_object *map,
             }
         }
 
+        /* Lookup raw_log_key */
+        if (ctx->raw_log_key && !raw_key && val.type == MSGPACK_OBJECT_STR) {
+            if (key.via.str.size == ctx->raw_log_key_len &&
+                    strncmp(key.via.str.ptr, ctx->raw_log_key, ctx->raw_log_key_len) == 0) {
+                raw_key = (char *) val.via.str.ptr;
```
Comment (edsiper) [inline plugins/out_kafka/kafka.c:220]: this routine is inside a loop, if I am not wrong the values might be set for one message but if the next message does not match they will have the value from the previous one (cc: @cosmo0920 )

**Review by edsiper — CHANGES_REQUESTED** [review]

File: plugins/out_kafka/kafka.c (hunk 2 above)
Comment (cosmo0920) [reply plugins/out_kafka/kafka.c:220]: Ohh, I've overlooked one. Thanks for catching this.

Comment (zecke) [reply plugins/out_kafka/kafka.c:220]: How is this different to the `message_key` handling a few lines of above? What is the lifetime of val.via.str.ptr? Shouldn't it be valid until after the msgpack is freed?

Comment (cosmo0920) [reply plugins/out_kafka/kafka.c:220]: Normally, we preserve the temporary reference as a heap memory with `flb_sds_create_len`.
In this case, just referencing via `val.via.str.ptr` would be replaced with the latter line in the loop.
So, we need to create a heap region with:


```c
 raw_key = flb_sds_create_len(val.via.str.ptr, val.via.str.size);
```

and in the finalizing function:

```c
flb_sds_destroy(raw_key);
```

Plus, this heap should be used `flb_sds_t` type for it.

Comment (zecke) [reply plugins/out_kafka/kafka.c]: Took a copy of that memory and freed it on the exit paths. flb_sds_destroy(NULL) seems legit (like free(NULL))

**Pull request merged**